File: ScalaOverview/src/main/scala/ISort.scala
object ISort {
  // Functions may be defined in any order, but must belong to an object
  def main(args: Array[String]): Unit = {
    val data = Array(3, 1, 4, 1, 5, 9, 2, 6, 5)
    isort(data)
    display(data)
  }
  
  def isort(a: Array[Int]): Unit = {
    for (i <- 1 until a.length) {
      val value = a(i)
      var j = i
      while (j > 0 && a(j-1) > value) {
        a(j) = a(j-1)
        j -= 1
      }
      a(j) = value
    }
  }
  
  def display(a: Array[Int]): Unit = {
    for (i <- 0 until a.length) {
      print(a(i) + " ")
    }
    println()
  }
}

File: ScalaOverview/src/main/cpp/ISort.cpp
#include <iostream>
using namespace std;

// Forward declarations of functions defined later
void isort(int a[], int N);
void display(int a[], int N);

int main()
{
  int data[] = {3, 1, 4, 1, 5, 9, 2, 6, 5};
  const int size = sizeof(data) / sizeof(int);

  isort(data, size);
  display(data, size);
}

void isort(int a[], int N)
{
  for (int i = 1; i < N; ++i) {
    int value = a[i];
    int j = i;
    while (j > 0 && a[j-1] > value) {
      a[j] = a[j-1];
      --j;
    }
    a[j] = value;
  }
}

void display(int a[], int N)
{
  for (int i = 0; i < N; ++i) {
    cout << a[i] << " ";
  }
  cout << endl;
}
